bin/database.py

```diff
@@ -51,6 +53,44 @@ def confirm_schema_match():
     else:
         sys.exit(0)
 
+def getMonotonicTime():
+    # http://stackoverflow.com/a/7424304
+    return os.times()[4]
+
+def process_cursor(cursor, closure):
+    """
+    Given an iterable (say, a mongo cursor) and a closure, call that closure in parallel over the iterable.
+    Call order is undefinied. Currently launches N python process workers, where N is the number of vcpu cores.
+
+    Useful for upgrades that need to touch each document in a database, and don't need an iteration order.
+    """
+
+    begin = getMonotonicTime()
+
+    cores = multiprocessing.cpu_count()
+    pool = multiprocessing.Pool(cores)
+    logging.info('Iterating over cursor with ' + str(cores) + ' workers')
+
+    # Launch all work, iterating over the cursor
+    # Note that this creates an array of n multiprocessing.pool.AsyncResults, where N is table size.
+    # Memory usage concern in the future? Doesn't seem to be an issue with ~120K records.
+    # Could be upgraded later with some yield trickery.
+    results = [pool.apply_async(closure, (document,)) for document in cursor]
+
+    # Read the results back, presumably in order!
+    for res in results:
+        result = res.get()
+        if result != True:
+            logging.info('Upgrade failed: ' + str(result))
```

Comment:
If an upgrade doesn't succeed isn't that a level of failure that should fail the upgrade?